packages/medusa/src/services/batch-job.ts

```diff
@@ -71,6 +144,220 @@ class BatchJobService extends TransactionBaseService<BatchJobService> {
+
+      let batchJob = await this.retrieve(batchJobId)
+
+      const { context, ...rest } = data
+      if (context) {
+        batchJob.context = await this.validateBatchContext(
+          batchJob.type,
+          context
+        )
+      }
+
+      Object.keys(rest)
+        .filter((key) => typeof rest[key] !== `undefined`)
+        .forEach((key) => {
+          batchJob[key] = rest[key]
+        })
+
+      batchJob = await batchJobRepo.save(batchJob)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.UPDATED, {
+          id: batchJob.id,
+        })
+
+      return batchJob
+    })
+  }
+
+  protected async updateStatus(
+    batchJobOrId: BatchJob | string,
+    status: BatchJobStatus
+  ): Promise<BatchJob | never> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        let batchJob: BatchJob = batchJobOrId as BatchJob
+        if (typeof batchJobOrId === "string") {
+          batchJob = await this.retrieve(batchJobOrId)
+        }
+
+        const { entityColumnName, eventType } =
+          this.batchJobStatusMapToProps.get(status) || {}
+
+        if (!entityColumnName || !eventType) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Unable to update the batch job status from ${batchJob.status} to ${status}. The status doesn't exist`
+          )
+        }
+
+        batchJob[entityColumnName] = new Date()
+
+        const batchJobRepo = transactionManager.getCustomRepository(
+          this.batchJobRepository_
+        )
+        batchJob = await batchJobRepo.save(batchJob)
+
+        this.eventBus_.withTransaction(transactionManager).emit(eventType, {
+          id: batchJob.id,
+        })
+
+        return batchJob
+      }
+    )
+  }
+
+  async confirm(batchJobOrId: string | BatchJob): Promise<BatchJob | never> {
+    return await this.atomicPhase_(async () => {
+      let batchJob: BatchJob = batchJobOrId as BatchJob
+      if (typeof batchJobOrId === "string") {
+        batchJob = await this.retrieve(batchJobOrId)
+      }
+
+      if (!batchJob.dry_run) {
+        return batchJob
+      }
+
+      if (batchJob.status !== BatchJobStatus.AWAITING_CONFIRMATION) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          "Cannot confirm a batch job that is not awaiting for confirmation"
+        )
+      }
+
+      return await this.updateStatus(batchJob, BatchJobStatus.CONFIRMED)
+    })
+  }
+
+  async complete(batchJobOrId: string | BatchJob): Promise<BatchJob | never> {
+    return await this.atomicPhase_(async () => {
+      let batchJob: BatchJob = batchJobOrId as BatchJob
+      if (typeof batchJobOrId === "string") {
+        batchJob = await this.retrieve(batchJobOrId)
+      }
+
+      if (
+        (batchJob.dry_run && batchJob.status !== BatchJobStatus.CONFIRMED) ||
+        (!batchJob.dry_run && batchJob.status !== BatchJobStatus.PROCESSING)
```

Comment:
**discussion**: 
I'm still on the fence regarding the event ordering we have. It doesn't make sense for one job to be in the "processing" state before completion if it is a non `dry_run` job and another to be in a "confirmed" state before completion if it is a `dry_run` job (since in both cases the jobs are actually in a state of processing). This is introducing confusion in the codebase and the mental model around batch jobs and I think we should "align" event order to be equal in both cases after "confirmed", wdyt?

**proposal**:
I have a proposal for the event ordering which requires a smaller number of states and could potentially simplify things. The main idea would be to move `awaiting_confirmation` before `ready` state. Also, we would consider "processing" to be only batch reading/writing from/to a DB and e.g. parsing a CSV file would be considered as preparation for a job.

1. **Created**
    - (**Awaiting confirmation)**
        - *this is just skipped in the case of non “dry_run” job*
        - *upon confirmation, the job is directly moved to the “ready” state*
        - *purpose of `prepareBatchJobForProcessing` should be to move a batch job to a consistent state*
          - _each strategy would define its own preparation work in this method_
          - _the method could coordinate other asynchronous operations_
2. **Ready**
    - *at this point, all jobs have the same event order and are handled the same way from this point forward*
3. **Processing**
    - “*processing” only refers to actual reading and writing from/to DB*
    - *although parsing a CSV file could be a lengthy operation it should be considered as a preparation step for actual processing (which is writing to DB)*
4. **Completed** | **Canceled** | **Failed**

I think it is a bit simpler abstraction that covers all types of batch jobs we have, but let me know what you think? cc: @pKorsholm @olivermrbl @srindom

Reply:
just to clarify the existing before answering :)

at the moment the flow is the following:
1. **CREATED**
Preparing job to be processed
2. **READY**
At this step, the job is ready to be processed
3. **Processing**
- if the user choose to run it in dry_run mode, then the processing will not really create the import/export and only extract the appropriate data or something to give back to the user
- if the user choose the run it as normal (not dry_run), then the job is just processed as usual and the data are really written into the target before sending it back to the user
4. **AWAITING_CONFIRMATION**
This state is only true for dry_run job, it means the user must confirm the action based on the info we gave to him
5. **CONFIRMED**
This state is only true for dry_run job, it means the user has confirmed that he wants to proceed with the export/import base on the data we gave him
6. **COMPLETED**
Means that since the user confirmed the action, we are just writing  the output where it must be written and mark the job as completed
7. **CANCELED**/*FAILED** for the rest

Reply:
yes these are the current events we have

Reply:
Which mean the awaiting confirmation is not before ready and that the order is always the same if you don’t look at the confirmation and the awaiting status

Reply:
Maybe a call could be good for that subject and then writing down the thoughts process? ☺️

Reply:
> the order is always the same if you don’t look at the confirmation and the awaiting status

This kinda proves my point and is why I'm proposing that we could move "awaiting_confirmation" before "ready" and remove "confirmed", redefine what each step means and have a more streamlined flow.

On the logical level, the order is almost the same in both cases but on the implementation level sometimes the actual import processing is done in "processing" (Step 5.) state and sometimes in "confirmed" (Step 3.) state (and this exception we have to keep track of while working).

But yes let's jump on a call and note down the conclusion :)

Reply:
I think that the model proposed is not descriptive enough when it comes to the lifetime of the batch job and how work is delegated between the core and the worker. We have up to two to three phases of work that is done on a batch job depending on wether or not its a dry_run. I will refer to these stages as such:

- Preparation
- Parsing (only relevant for dry_run)
- Processing

The first of these phases “Preparation” should be done in the core when after the batch job has been created. At this point the relevant files from the request can be uploaded and any other invariants can be established before the work on the batch job begins. It’s also at this step that the context is validated. We require some work to be done in the API before hand, but this step allows us to limit this as much as possible before handover to a worker can happen. Thus in the model you have described this step would happen before the “Created”. In the current event model this happens after "Created", and when preparation is done we emit “Ready”.

At this point a worker should be able to take over the batch job and enter the “Parsing” state. At this state whatever work is required for the batch job to enter the “awaiting confirmation” state should be performed. This would happen between the “created” and the “awaiting confirmation” state in your model.

Whenever the batch job has been confirmed, or has been parsed we should go to the “Processing” state, which involves writing to the database as you’ve described. 

I don’t think we can combine the “created” and the “ready” in the current event model, since the “created” event is emitted on creation and some work and preparation of the batch job can still be required before any handoff where a worker can begin the “Parsing” phase.

As far as i can tell there is a difference between the two event models of length one and it’s between created and awaiting confirmation. I believe we need an event here in order to create a separation between Preparation and Parsing and to delegate the Parsing phase to a worker.

**Suggestion:** 
We could rename the current “Processing” to “Preparing”, and “Confirmed” to “Processing”. In that way the state journeies of batch jobs would be: 

dry_run: 
> Created → Ready → Preparing → Awaiting Confirmation → Processing → Complete

non dry_run: 
> Created → Ready → Processing → Complete

It's only semantic but would prevent any work to be done during a step called "Confirmed" which sounds like the batch job is at rest. In this way we convey that work is actively being done on the batch job.

Reply:
I think we will create a call to gather all the info and make a proper description of the status to be used and there purpose 🚀

Reply:
After a discussion with @fPolic and @pKorsholm, here is our thoughts process on the batch job status and handling:

Let's take an example to settle that down mentally, let's go for the export (but not limited to):
-> The user ask for an export
-> CREATED the batchJob is being created and move to PRE-PROCESSING  
-> PRE-PROCESSING the worker pick up the job to launch the pre processing process and marked the job as READY
-> READY
--> --> if the job is in DRY_RUN mode then we are stopping there until a user call the api end point to emit the PROCESSING event
--> --> if the job is not in DRY_RUN mode then the job will immediately be marked as PROCESSING
-> PROCESSING the worker pick up the job to launch the processing process and marked the job as COMPLETE

RETRY

When the job is being processed, some errors can occurs during that process.
The CONTEXT will hold a property max_retry that will hold the number of time a job can be processed
If an error is emitted by ourselves, then we know the job should (maybe) not be retried, in the case we can set the FAILED_AT value on the job
if an error occurs outside of our hands, then we can increase the retry_count property on the context to know if the batch job should be retried or not depending on if it has reach the max already or not.